controllers/agentpool_controller_autoscaling.go

```diff
@@ -12,115 +12,98 @@ import (
-}
 
-func getTargetWorkspaceIDsByWildcardName(ctx context.Context, ap *agentPoolInstance, targetWorkspace appv1alpha2.TargetWorkspace) ([]string, error) {
-	list, err := ap.tfClient.Client.Workspaces.List(ctx, ap.instance.Spec.Organization, &tfc.WorkspaceListOptions{
-		WildcardName: targetWorkspace.WildcardName,
-	})
-	if err != nil {
-		return []string{}, err
-	}
-	workspaceIDs := []string{}
-	for _, w := range list.Items {
-		workspaceIDs = append(workspaceIDs, w.ID)
+	if ap.instance.Spec.AgentDeploymentAutoscaling.TargetWorkspaces == nil {
+		return int32(len(workspaceNames)), nil
 	}
-	return workspaceIDs, nil
-}
 
-func computeRequiredAgents(ctx context.Context, ap *agentPoolInstance) (int32, error) {
-	required := 0
-	workspaceIDs, err := getTargetWorkspaceIDs(ctx, ap)
-	if err != nil {
-		return 0, err
-	}
-	for _, workspaceID := range workspaceIDs {
-		r, err := computeRequiredAgentsForWorkspace(ctx, ap, workspaceID)
-		if err != nil {
-			return 0, err
+	for _, t := range *ap.instance.Spec.AgentDeploymentAutoscaling.TargetWorkspaces {
+		switch {
+		case t.Name != "":
+			if _, ok := workspaceNames[t.Name]; ok {
+				required++
+				delete(workspaceNames, t.Name)
+			}
+		case t.ID != "":
+			if _, ok := workspaceIDs[t.ID]; ok {
+				required++
+			}
+		case t.WildcardName != "":
+			// This is not a mistake here.
+			// Both 'prefix' and 'suffix' indicate whether a part of the name is in the prefix, suffix, or both.
+			// If the wildcard indicator '*' is in the suffix part, then search for a substring that is in the prefix.
+			// If the wildcard indicator '*' is in the prefix part, then search for a substring that is in the suffix.
+			// If the wildcard indicator '*' is in both the prefix and the suffix, then search for a substring that is in between '*'.
+			// For example:
+			// (1) 'hcp-terraform-workspace-*' -- the wildcard indicator '*' is at the end of the wildcard name (suffix),
+			// therefore, we should search for a workspace name that starts with the prefix 'hcp-terraform-workspace-'.
+			// (2) '*-terraform-workspace' -- the wildcard indicator '*' is at the beginning of the wildcard name (prefix),
+			// therefore, we should search for a workspace name that ends with the suffix '-terraform-workspace'.
+			// (3) '*-terraform-workspace-*' -- the wildcard indicator '*' is at the beginning and the end of the wildcard name (prefix and suffix),
+			// therefore, we should search for a workspace name containing the substring '-terraform-workspace-'.
+			prefix := strings.HasSuffix(t.WildcardName, "*")
+			suffix := strings.HasPrefix(t.WildcardName, "*")
+			wn := strings.Trim(t.WildcardName, "*")
+			for w := range workspaceNames {
+				match := false
+				switch {
+				case prefix && suffix:
+					match = strings.Contains(w, wn)
+				case prefix:
+					match = strings.HasPrefix(w, wn)
+				case suffix:
+					match = strings.HasSuffix(w, wn)
+				}
+				if match {
+					required++
+					delete(workspaceNames, w)
+				}
+			}
```

Comment:
It's not great that we have to duplicate this logic from the TFC codebase but I see why it is like this. I would consider putting this logic into its own function and unit testing it, at least until we deprecate and remove the `wildcardName` field.